python/ray/util/sgd/v2/backends/backend.py

```diff
@@ -1,45 +1,128 @@
+
+    def get_backend_cls(self):
+        raise NotImplementedError
 
 
 class BackendExecutor:
+    """Main execution class for training backends.
+
+    This class holds a worker group and is responsible for executing the
+    training function on the workers, and collecting intermediate results
+    from ``sgd.report()``.
+
+    Args:
+        backend_config (BackendConfig): The configurations for this
+            specific backend.
+        num_workers (int): Number of workers to use for training.
+        num_cpus_per_worker (float): Number of CPUs to use per worker.
+        num_gpus_per_worker (float): Number of GPUs to use per worker.
+    """
+
     def __init__(self,
                  backend_config: BackendConfig,
                  num_workers: int = 1,
-                 num_cpus_per_worker: int = 1,
-                 num_gpus_per_worker: int = 0):
+                 num_cpus_per_worker: float = 1,
+                 num_gpus_per_worker: float = 0):
         self._backend_config = backend_config
+        self._backend = self._backend_config.get_backend_cls()
         self._num_workers = num_workers
         self._num_cpus_per_worker = num_cpus_per_worker
         self._num_gpus_per_worker = num_gpus_per_worker
 
-    def start(self):
+        self.worker_group = InactiveWorkerGroup()
+
+    def start(self, initialization_hook: Callable):
+        """Starts the worker group."""
         self.worker_group = WorkerGroup(self._num_workers,
                                         self._num_cpus_per_worker,
                                         self._num_gpus_per_worker)
+        self.worker_group.execute(initialization_hook)
+        self._backend.on_start()
+
+    def run(self, train_func: Callable[[], T]) -> List[T]:
+        """Executes training function on all workers and yield results.
+
+        Args:
+            train_func (Callable): The training function to run on each
+                worker. It must not have any required arguments.
+
+        Yields:
+            A list of values passed to ``sgd.report()`` from each worker.
+        """
+        # Run the training function asynchronously.
+        training_futures = self.worker_group.execute_async(train_func)
+
+        return self.get_with_failure_handling(training_futures)
+
+    def get_with_failure_handling(self, remote_values):
+        """Gets the remote values while handling for worker failures.
+
+        Args:
+            remote_values (list): List of object refs representing functions
+                that may fail in the middle of execution. For example, running
+                a SGD training loop in multiple parallel actor calls.
+
+        Returns:
+            The objects represented the passed in ObjectRefs.
+        """
+        unfinished = remote_values
+        try:
+            while len(unfinished) > 0:
+                finished, unfinished = ray.wait(unfinished)
+                ray.get(finished)
```

Comment:
Can you add a comment here explaining that `ray.get` is used to expose the  `RayActorError` when a failure occurs?

Reply:
```suggestion
                # If a failure occurs the ObjectRef will be marked as finished.
                # Calling ray.get will expose the failure as a RayActorError.
                ray.get(finished)
```